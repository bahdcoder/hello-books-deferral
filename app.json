{
  "name": "Hello Books",
  "description": "A system that allows efficient, cost-effective management of Books for Libraries."
}
