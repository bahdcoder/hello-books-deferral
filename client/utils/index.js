export const getDateStr = str => {
    const end = str.indexOf('T');
    const rawDate = str.slice(0, end);
    const firstHyphen = rawDate.indexOf('-');
    const lastHyphen = rawDate.lastIndexOf('-');
    const year = rawDate.slice(0, firstHyphen);
    const month = rawDate.slice(firstHyphen + 1, lastHyphen);
    const day = rawDate.slice(lastHyphen + 1);
    const monthStrArr = [
        'Jan',
        'Feb',
        'Mar',
        'Apr',
        'May',
        'Jun',
        'Jul',
        'Aug',
        'Sep',
        'Oct',
        'Nov',
        'Dec'
    ];
    const monthStr = monthStrArr[Number(month) - 1];
    return `${monthStr} ${day} ${year}`;
};
