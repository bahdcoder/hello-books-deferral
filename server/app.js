import express from 'express';
import consola from 'consola';
import bodyParser from 'body-parser';
import swagger from 'swagger-ui-express';
import Knex from 'knex';
import { createServer } from 'http';
import { Model } from 'objection';
import swaggerDocument from '@/swagger.json';
import Routes from '@routes/v1';
import config from '@config';

import 'express-jsend';

export const app = express();

export const databaseConnection = Knex(config.database[config.server.env]);

Model.knex(databaseConnection);

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));

Routes(app);

app.get('/', (req, res) => {
    res.status(200).jsend({ message: 'Hello Books Deferral' });
});
/**
 *  setup Swagger
 */
app.use(
    '/api-docs',
    swagger.serve,
    swagger.setup(swaggerDocument, { explorer: true })
);

export const server = createServer(app);

export default { app, databaseConnection };
