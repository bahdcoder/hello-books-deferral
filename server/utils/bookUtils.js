const { validations } = require("indicative");
const { Vanilla } = require("indicative/builds/formatters");
const Validator = require("indicative/builds/validator");
const Book = require("../models/Book");

/* custom user friendly error messages */
const messages = {
  required: "{{ field }} is required to create a book",
  string: "{{ field }} is not a string",
  number: "{{ field }} is not a number",
  different: "{{ field }} must be unique"
};

// Trim white spaces
const sanitizeRules = {
  title: "trim",
  coverType: "trim",
  description: "trim",
  isbn: "trim",
  publisher: "trim",
  year: "trim",
  copiesAvailable: "trim"
};

/* add the unique custom validator to indicative validations object */
validations.different = async (data, field, message, args, get) => {
  const value = get(data, field);
  if (!value) return;
  const column = args[1];
  const row = await Book.query().where(column, value);
  if (row[0]) throw message;
};

const validatorInstance = Validator(
  validations,
  Vanilla
); /* create custom validator */

module.exports = { validatorInstance, messages, sanitizeRules };
