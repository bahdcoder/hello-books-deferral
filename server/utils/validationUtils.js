import { validations } from "indicative";
import { Vanilla } from "indicative/builds/formatters";
import Validator from "indicative/builds/validator";
import User from "@models/User";

/* custom user friendly error messages */
export const messages = {
  required: "{{ field }} is required to create a new account",
  string: "{{ field }} is not a string",
  unique: "{{ field }} must be unique",
  email: "{{ field }} is invalid",
  min: "{{ field }} is must be less than {{ argument.0 }}",
  alpha_numeric: "only numbers and letters are allowed for {{ field }}",
  confirmed: "Confirm {{ field }} does not match {{ field }}"
};

/* sanitization rule to trim whitespaces */
export const sanitizeRules = {
  firstName: "trim",
  lastName: "trim",
  email: "trim",
  password: "trim"
};

/* add the unique custom validator to indicative validations object */
validations.unique = async (data, field, message, args, get) => {
  const value = get(data, field);
  if (!value) return;
  const column = args[1];
  /* check user table to see if there is an existing user email */
  const row = await User.query().where(column, value);
  if (row[0]) throw message;
};

export const validatorInstance = Validator(
  validations,
  Vanilla
); /* create custom validator */
