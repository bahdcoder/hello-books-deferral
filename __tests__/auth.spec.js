import supertest from 'supertest';
import { app, databaseConnection } from '@server/app';
import { getUser, createUser } from '@tests/utils/helpers';

const server = () => supertest(app);

describe('AUTH API ENDPOINTS', () => {
    beforeAll(async () => {
        await databaseConnection('users').truncate();
        await databaseConnection.migrate.latest();
    });

    describe('POST SIGN UP api/v1/auth/signup', () => {
        it('should not sign up user if email is empty', async () => {
            const { status, body } = await server()
                .post('/api/v1/auth/signup')
                .send({
                    ...getUser(),
                    email: ''
                });

            expect(status).toBe(422);
            expect(body).toMatchSnapshot();
        });

        it('should sign up user with valid inputs', async () => {
            const user = getUser();
            const { status, body } = await server()
                .post('/api/v1/auth/signup')
                .send(user);

            expect(status).toBe(201);
            expect(body.data.user.email).toBe(user.email);
            expect(Object.keys(body.data)).toMatchSnapshot();
        });

        it('should not sign up user that is already registered', async () => {
            const user = getUser();
            await createUser(user);

            const { status, body } = await server()
                .post('/api/v1/auth/signup')
                .send(user);

            expect(status).toBe(422);
            expect(body).toMatchSnapshot();
        });

        it('should not verify an email with a wrong verification code', async () => {
            const { status, body } = await server()
                .post('/api/v1/auth/verify/:thisistherandomstring')
                .send();

            expect(status).toBe(400);
            expect(body).toMatchSnapshot();
        });

        it('should verify an email with a good verification code', async () => {
            const user = getUser();
            const { email_confirm_code } = await createUser(user);

            const { status, body } = await server().post(
                `/api/v1/auth/verify/${email_confirm_code}`
            );

            expect(status).toBe(200);
            expect(body).toMatchSnapshot();
        });
    });
});
