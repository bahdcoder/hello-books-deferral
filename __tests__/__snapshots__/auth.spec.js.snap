// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AUTH API ENDPOINTS PATCH PASSWORD RESET api/v1/auth/reset should not reset password if password is invalid 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "password",
      "message": "password must be more than 8",
      "validation": "min",
    },
  ],
  "status": "error",
}
`;

exports[`AUTH API ENDPOINTS PATCH PASSWORD RESET api/v1/auth/reset should not reset password if token is invalid 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "message": "reset token is invalid",
    },
  ],
  "status": "error",
}
`;

exports[`AUTH API ENDPOINTS PATCH PASSWORD RESET api/v1/auth/reset should reset password if all inputs are valid 1`] = `
Object {
  "data": Object {
    "message": "password reset successfull",
  },
  "status": "success",
}
`;

exports[`AUTH API ENDPOINTS POST CREATE USER api/v1/auth/create-user A super-admin should be able to create a normal user 1`] = `
Object {
  "data": Object {
    "message": "user has been created and mail sent to user",
  },
  "status": "success",
}
`;

exports[`AUTH API ENDPOINTS POST LOGIN api/v1/auth/login should reject login if email is not found  1`] = `
Object {
  "code": Object {
    "error": "Email or password is invalid.",
  },
  "status": "error",
}
`;

exports[`AUTH API ENDPOINTS POST LOGIN api/v1/auth/login should reject login if password is wrong  1`] = `
Object {
  "code": Object {
    "error": "Email or password is invalid.",
  },
  "status": "error",
}
`;

exports[`AUTH API ENDPOINTS POST PASSWORD RESET api/v1/auth/reset should not send mail if user email is invalid 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "email",
      "message": "email is invalid",
      "validation": "email",
    },
  ],
  "status": "error",
}
`;

exports[`AUTH API ENDPOINTS POST PASSWORD RESET api/v1/auth/reset should not send mail if user is not found 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "message": "email does not exist",
    },
  ],
  "status": "error",
}
`;

exports[`AUTH API ENDPOINTS POST PASSWORD RESET api/v1/auth/reset should send email if user is valid 1`] = `
Array [
  "message",
]
`;

exports[`AUTH API ENDPOINTS POST SIGN UP api/v1/auth/signup should not sign up user if email is empty 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "email",
      "message": "email is required",
      "validation": "required",
    },
  ],
  "status": "error",
}
`;

exports[`AUTH API ENDPOINTS POST SIGN UP api/v1/auth/signup should not sign up user that is already registered 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "email",
      "message": "email must be unique",
      "validation": "unique",
    },
  ],
  "status": "error",
}
`;

exports[`AUTH API ENDPOINTS POST SIGN UP api/v1/auth/signup should not verify an email with a wrong verification code 1`] = `
Object {
  "code": Object {
    "error": "Bad Request",
  },
  "status": "error",
}
`;

exports[`AUTH API ENDPOINTS POST SIGN UP api/v1/auth/signup should sign up user with valid inputs 1`] = `
Array [
  "message",
  "token",
  "user",
]
`;

exports[`AUTH API ENDPOINTS POST SIGN UP api/v1/auth/signup should verify an email with a good verification code 1`] = `
Object {
  "data": Object {
    "message": "User verified",
  },
  "status": "success",
}
`;
